File: leetcode/802.py
class Solution:
    def eventualSafeNodes(self, graph_: List[List[int]]) -> List[int]:
        n: int = len(graph_)
        graph: List[Set[int]] = [set(graph_[i]) for i in range(n)]
        reverse_adj_lists: List[List[int]] = [list() for _ in range(n)]
        for i in range(n):
            for j in graph[i]:
                reverse_adj_lists[j].append(i)
        ans = [i for i in range(n) if len(graph[i]) == 0]
        st = ans.copy()
        while len(st):
            cur = st.pop()
            for nex in reverse_adj_lists[cur]:
                graph[nex].remove(cur)
                if len(graph[nex]) == 0:
                    st.append(nex)
                    ans.append(nex)
        return sorted(ans)

File: leetcode/802.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        queue<int> q;
        const int n = graph.size();
        vector<int> edge_counts(n, 0);
        vector<vector<int>> rev_graph(n, vector<int>());
        for (int i = 0; i < graph.size(); ++i) {
            if (graph[i].size() == 0)
                q.push(i);
            for (const auto& nex : graph[i]) {
                rev_graph[nex].push_back(i);
                ++edge_counts[i];
            }
        }
        vector<int> safe_nodes;
        while (q.size()) {
            const auto cur = q.front();
            q.pop();
            safe_nodes.push_back(cur);
            for (const auto& prev : rev_graph[cur]) {
                --edge_counts[prev];
                if (edge_counts[prev] == 0)
                    q.push(prev);
            }
        }
        sort(safe_nodes.begin(), safe_nodes.end());
        return safe_nodes;
    }
};
